language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -187,16 +204,21 @@ dummyType = TyVar () dummyTyName
 
 -- | Extract type of a term. The resulting type is normalized.
 typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (NormalizedType TyNameWithKind ())
-typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = pure (void $ NormalizedType ty) -- annotations on types must be normalized
-typeOf (LamAbs _ _ ty t)                         = NormalizedType <$> (TyFun () (void ty) <$> (getNormalizedType <$> typeOf t))
+typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = do
+    (norm, _) <- ask
+    maybeRed norm (void ty)
+typeOf (LamAbs _ _ ty t)                         = do
+    (norm, _) <- ask
+    (NormalizedType ty') <- maybeRed norm (void ty)
+    NormalizedType <$> (TyFun () ty' <$> (getNormalizedType <$> typeOf t))
 typeOf (Error x ty)                              = do
     k <- kindOf ty
     case k of
-        Type{} -> pure (void $ NormalizedType ty) -- annotations on types must be normalized
+        Type{} -> pure (void $ NormalizedType ty)
```

Comment:
We require `ty` in `Error x ty` to be normalized?

Reply:
Yes, that is correct.

Reply:
Oh you're right. The checking in `Language.PlutusCore.Normalize` diverges from the spec.

I believe this is a problem in the spec, as we would like type synthesis to produce normalized types.

Reply:
> I believe this is a problem in the spec, as we would like type synthesis to produce normalized types.

I'll ask Rebecca.

Reply:
In any case I think we should allow instantiating `error` at a non-normalized type.

Reply:
It should be `Type{} -> pure (tyReduce $ void ty)` as per our consensus about allowing `error` to be instantiated at a non-normalized type.